controller/src/main/java/org/jboss/as/controller/CapabilityRegistry.java

```diff
@@ -0,0 +1,745 @@
+     */
+    @Override
+    public void registerCapability(RuntimeCapabilityRegistration capabilityRegistration) {
+        writeLock.lock();
+        try {
+            CapabilityId capabilityId = capabilityRegistration.getCapabilityId();
+            RegistrationPoint rp = capabilityRegistration.getOldestRegistrationPoint();
+            RuntimeCapabilityRegistration currentRegistration = capabilities.get(capabilityId);
+            if (currentRegistration != null) {
+                // The actual capability must be the same, and we must not already have a registration
+                // from this resource
+                if (!Objects.equals(capabilityRegistration.getCapability(), currentRegistration.getCapability())
+                        || !currentRegistration.addRegistrationPoint(rp)) {
+                    throw ControllerLogger.MGMT_OP_LOGGER.capabilityAlreadyRegisteredInContext(capabilityId.getName(),
+                            capabilityId.getContext().getName());
+                }
+                // else it was ok, and we just recorded the additional registration point
+            } else {
+                capabilities.put(capabilityId, capabilityRegistration);
+            }
+
+            // Add any hard requirements
+            for (String req : capabilityRegistration.getCapability().getRequirements()) {
+                registerRequirement(new RuntimeRequirementRegistration(req, capabilityId.getName(),
+                        capabilityId.getContext(), rp));
+            }
+
+            if (!forServer) {
+                CapabilityContext capContext = capabilityId.getContext();
+                knownContexts.add(capContext);
+            }
+            modified = true;
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Registers an additional requirement a capability has beyond what it was aware of when {@code capability}
+     * was passed to {@link #registerCapability(RuntimeCapabilityRegistration)}. Used for cases
+     * where a capability optionally depends on another capability, and whether or not that requirement is needed is
+     * not known when the capability is first registered.
+     *
+     * @param requirement the requirement
+     * @throws java.lang.IllegalArgumentException if no matching capability is currently
+     *                                            {@link #registerCapability(RuntimeCapabilityRegistration) registered} for either {@code required} or {@code dependent}
+     */
+    @Override
+    public void registerAdditionalCapabilityRequirement(RuntimeRequirementRegistration requirement) {
+        writeLock.lock();
+        try {
+            registerRequirement(requirement);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    private void registerRequirement(RuntimeRequirementRegistration requirement) {
+        try {
+            writeLock.lock();
```

Comment:
This should be outside the try if the release is in the finally.

But you don't need to lock/unlock in this method, as it is called with the lock held. You can document it as "must be called with the write lock held".